kafka/src/main/java/org/axonframework/extensions/kafka/eventhandling/cloudevent/CloudEventKafkaMessageConverter.java

```diff
@@ -0,0 +1,463 @@
+        this.sequencingPolicy = builder.sequencingPolicy;
+        this.upcasterChain = builder.upcasterChain;
+        this.extensionNameResolver = builder.metadataToExtensionMap;
+        this.metadataNameResolver = builder.metadataToExtensionMap.entrySet()
+                                                                  .stream()
+                                                                  .collect(Collectors.toMap(Map.Entry::getValue,
+                                                                                            Map.Entry::getKey));
+        this.sourceSupplier = builder.sourceSupplier;
+        this.dataContentTypeSupplier = builder.dataContentTypeSupplier;
+        this.dataSchemaSupplier = builder.dataSchemaSupplier;
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link CloudEventKafkaMessageConverter}.
+     * <p>
+     * The {@link SequencingPolicy} is defaulted to an {@link SequentialPerAggregatePolicy}. The {@link Serializer} is
+     * a
+     * <b>hard requirement</b> and as such should be provided.
+     *
+     * @return a Builder to be able to create a {@link CloudEventKafkaMessageConverter}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Note that the {@link ProducerRecord} created through this method sets the {@link ProducerRecord#timestamp()} to
+     * {@code null}. Doing so will ensure the used Producer sets a timestamp itself for the record. The
+     * {@link EventMessage#getTimestamp()} field is however still taken into account, but as via the CloudEvent.
+     * <p>
+     * Additional note that the ProducerRecord will be given a {@code null} {@link ProducerRecord#partition()} value. In
+     * return, the {@link ProducerRecord#key()} field is defined by using the configured {@link SequencingPolicy} to
+     * retrieve the given {@code eventMessage}'s {@code sequenceIdentifier}. The combination of a {@code null} partition
+     * and the possibly present or empty key will define which partition the Producer will choose to dispatch the record
+     * on.
+     *
+     * @see ProducerRecord
+     */
+    @Override
+    public ProducerRecord<String, CloudEvent> createKafkaMessage(EventMessage<?> eventMessage, String topic) {
+        SerializedObject<byte[]> serializedObject = eventMessage.serializePayload(serializer, byte[].class);
+        return new ProducerRecord<>(
+                topic, null, null, recordKey(eventMessage),
+                toCloudEvent(eventMessage, serializedObject),
+                null
+        );
+    }
+
+    private CloudEvent toCloudEvent(EventMessage<?> message, SerializedObject<byte[]> serializedObject) {
+        CloudEventBuilder builder = new CloudEventBuilder();
+        builder.withId(message.getIdentifier());
+        builder.withData(serializedObject.getData());
+        dataContentTypeSupplier.apply(message).ifPresent(builder::withDataContentType);
```

Comment:
Can we detect the format (JSON or XML) here?

Reply:
Hmm, seems not, unless we check the type of the eventSerializer :thinking:

Reply:
We could do something smart depending on the serializer used. But I'm not sure that is a good idea.

Reply:
Why would you want that, @Morlack? Does it provide direct/necessary benefit?

Reply:
It's one less thing to set, but since it's not mandatory I rather not set it. Even if we could check the serializer and set it to json for Jackson, and XML when used with XStream.

Reply:
@smcvb The events have to be deserialized at the client, would be nice to know the format (I think). However, setting it to a value depending on the serializer might not be great as well; the Serializer might not even be used and correspond with the wrong format.
So since it's optional, we'd probably better skip it.